waiter/src/waiter/scheduler.clj

```diff
@@ -1251,3 +1212,168 @@
+   last-request-time. We then build a set of corresponding source tokens to the service-ids and see if they point to a
+   service-id that is not running. We attempt to start a new service for those tokens/service-ids (not including
+   run-as-requester or parameterized services).
+
+   Returns a map with keys:
+   :exit-chan when a message is put on the exit-chan, the daemon terminates
+   :go-chan the corresponding async/go chan of the daemon. This can be used to see if the daemon has terminated.
+   :query-chan channel that puts the state in a response-chan after the daemon processes other channels first (e.g timer-chan)
+   :query-state-fn function that queries the current state of the daemon"
+  [clock timer-chan service-id->source-tokens-entries-fn service-id->metrics-fn service-id->stale-info start-new-service-fn
+   retrieve-descriptor-fn fallback-state-atom kv-store]
+  (cid/with-correlation-id
+    "start-new-services-maintainer"
+    (let [correlation-id (cid/get-correlation-id)
+          exit-chan (async/promise-chan)
+          query-chan (async/chan)
+          state-atom (atom {:last-update-time nil
+                            :service-id->last-request-time {}})
+          query-state-fn (fn query-state-fn [_]
+                           (assoc @state-atom :supported-include-params []))
+          go-chan
+          (async/go
+            (try
+              (loop [{:keys [service-id->last-request-time] :as current-state} @state-atom]
+                (reset! state-atom current-state)
+                (let [[msg current-chan]
+                      (async/alts! [exit-chan timer-chan query-chan] :priority true)
+                      next-state
+                      (condp = current-chan
+                        exit-chan
+                        (do
+                          (log/warn "stopping start-new-services-maintainer")
+                          (when (not= :exit msg)
+                            (throw (ex-info "Stopping start-new-services-maintainer" {:time (clock) :reason msg}))))
+
+                        timer-chan
+                        (timers/start-stop-time!
+                          (metrics/waiter-timer "core" "start-new-services-maintainer" "process-services")
+                          (try
+                            (let [fallback-state @fallback-state-atom
+                                  new-service-id->last-request-time
+                                  (create-service-id->last-request-time-from-metrics service-id->metrics-fn)
+                                  tokens-with-new-last-request-times
+                                  (get-tokens-to-start-services-for
+                                    correlation-id service-id->stale-info service-id->source-tokens-entries-fn kv-store
+                                    new-service-id->last-request-time service-id->last-request-time)]
+                              (doseq [token tokens-with-new-last-request-times]
```

Comment:
Let's move this to a helper function. It makes the go-block smaller and also allows unit testing the function.